### PR TITLE
15022 Similar Owner Names

### PR DESCRIPTION
**Pull request by owillborn**
*Issue #:* /bcgov/entity#15022

*Description of Changes*
- Search results with multiple copies are collapsed and the number and type of registrations are displayed
image

- Selected search results with matching registration numbers are collapsed on review screen as follows. The numbers beside the owner names represent the sum of active, exempt, and historical results under that name.
image

- Fixed fee summary to display fee according to unique registrations
image

- Other minor UI changes: Checkbox alignment, title text, Icon on selection screen, other text updates, tooltip on registration numbers with multiple owner names displayed.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the PPR license (Apache 2.0).

### CODE REVIEW

**Comment by cameron-eyds** [comment]
@owillborn ~Small comment here, if this is still a work in progress you can set the PR to a DRAFT state.~
Never mind, looks like you've done that :)

**Review by ChasBoucher — APPROVED** [review]
Looks good

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue [hunk 1]
```diff
@@ -383,20 +428,52 @@ export default defineComponent({
...
+    const hasMultipleStatus = (item: ManufacturedHomeSearchResultIF): boolean => {
+      return (item.activeCount > 0 && item.exemptCount > 0) ||
+             (item.activeCount > 0 && item.historicalCount > 0) ||
+             (item.exemptCount > 0 && item.historicalCount > 0)
+    }
+
+    const noSelectedOwner = (item: ManufacturedHomeSearchResultIF): boolean => {
+      var filteredResults = localState.results?.filter(result => result.mhrNumber === item.mhrNumber)
+      filteredResults = filteredResults.filter(result => result.selected === true)
+      console.log(filteredResults)
```
Comment (owillborn) [inline ppr-ui/src/components/tables/mhr/SearchedResults.vue]: Noted to be removed after review

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue [hunk 2]
```diff
@@ -358,6 +387,11 @@ export default defineComponent({
       router.push({ name: RouteNames.MHRSEARCH_CONFIRM })
     }
 
+    const hasSimilarOwners = (item: ManufacturedHomeSearchResultIF): boolean => {
```
Comment (dimak1) [inline ppr-ui/src/components/tables/mhr/SearchedResults.vue]: It's actually the same results, not similar. I would rename it to `hasMultipleSelections`.

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue [hunk 3]
```diff
@@ -369,8 +403,19 @@ export default defineComponent({
...
     }
 
+    const getOwnerCount = (item: ManufacturedHomeSearchResultIF): string => {
+      const count = item.activeCount + item.exemptCount + item.historicalCount
+      if (count > 1) return `(${count})`
+    }
     const getItemClass = (item: ManufacturedHomeSearchResultIF): string => {
-      return item.selected && !props.isReviewMode ? 'selected' : ''
+      var rowClass = ''
+      if (props.isReviewMode && localState.uniqueResults < localState.results) {
```
Comment (dimak1) [inline ppr-ui/src/components/tables/mhr/SearchedResults.vue]: Why do we need to check if unique results are less than actual results?

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue [hunk 4]
```diff
@@ -383,20 +428,52 @@ export default defineComponent({
...
         else return ownerStatus
       } else return ''
     }
 
+    // return adaptive text for owner status count(s)
+    const getOwnerStatusText = (item: ManufacturedHomeSearchResultIF): string => {
+      var returnText = ''
+      if (item.activeCount > 0) {
+        returnText += 'ACTIVE'
+        if (item.activeCount > 1) returnText += ' (' + item.activeCount + ')'
```
Comment (dimak1) [inline ppr-ui/src/components/tables/mhr/SearchedResults.vue]: There must be a better approach to re-write it. Maybe try string templating?

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue [hunk 5]
```diff
@@ -406,6 +483,10 @@ export default defineComponent({
         return result.includeLienInfo !== true ? { ...result, includeLienInfo: false } : result
       })
 
+      // sort search results
+      const sortedResults = orderBy(localState.results, ['mhrNumber', 'ownerName.lastName', 'ownerName.middleName', 'ownerName.firsName'], ['asc', 'asc', 'asc', 'asc'])
```
Comment (dimak1) [inline ppr-ui/src/components/tables/mhr/SearchedResults.vue]: Why do we need to sort them? Isn't the backend sorting it for us?

**Review by dimak1 — APPROVED** [review]
Overall looks good, nice job! Just few comments. Also would need to merge with my changes and simplify that css 😆

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue [hunk 6]
```diff
@@ -40,7 +41,13 @@
       </v-row>
       <v-row v-else class="result-info">
         <v-col id="review-results-count" cols="auto">
-          <span><b>{{ selectedMatchesLength }}</b> Manufactured Homes</span>
+          <span>Matches Selected: <b>{{ selectedMatchesLength }}</b></span>
+        </v-col>
+        <v-col id="review-registrations-count" cols="auto">
+          <span>Registrations:  <b>{{ uniqueResults.length }}</b></span>
```
Comment (cameron-eyds) [inline ppr-ui/src/components/tables/mhr/SearchedResults.vue:47]: What does `uniqueResults.length` resolve too if uniqueResults is null or undefined?

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue [hunk 7]
```diff
@@ -369,8 +403,19 @@ export default defineComponent({
       } else return '-'
     }
 
+    const getOwnerCount = (item: ManufacturedHomeSearchResultIF): string => {
+      const count = item.activeCount + item.exemptCount + item.historicalCount
+      if (count > 1) return `(${count})`
+    }
     const getItemClass = (item: ManufacturedHomeSearchResultIF): string => {
-      return item.selected && !props.isReviewMode ? 'selected' : ''
+      var rowClass = ''
```
Comment (cameron-eyds) [inline ppr-ui/src/components/tables/mhr/SearchedResults.vue]: Try and use let to keep your variables scope to it's wrapping { }. 
In this case, it's the same as var but it's best practices. 

var - scoped to the parent function
let - scoped to the wrapping { } 

More granular control over your variables scope. Js can do funny things when the global scopes becoming polluted with all sorts of variables and references 👍

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue [hunk 8]
```diff
@@ -383,20 +428,52 @@ export default defineComponent({
...
-      if (ownerStatus) {
+      if (ownerStatus === 'PREVIOUS') {
         if (ownerStatus === 'PREVIOUS') return 'HISTORICAL'
         else return ownerStatus
       } else return ''
     }
 
+    // return adaptive text for owner status count(s)
+    const getOwnerStatusText = (item: ManufacturedHomeSearchResultIF): string => {
+      var returnText = ''
```
Comment (cameron-eyds) [inline ppr-ui/src/components/tables/mhr/SearchedResults.vue]: same as above

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue (hunk 4 above)
Comment (cameron-eyds) [reply ppr-ui/src/components/tables/mhr/SearchedResults.vue]: An example: `returnText += (${item.activeCount})`

Comment (cameron-eyds) [reply ppr-ui/src/components/tables/mhr/SearchedResults.vue]: Just need to wrap your `${}` in back ticks

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue [hunk 9]
```diff
@@ -383,20 +428,52 @@ export default defineComponent({
...
+
+    const hasMultipleStatus = (item: ManufacturedHomeSearchResultIF): boolean => {
+      return (item.activeCount > 0 && item.exemptCount > 0) ||
+             (item.activeCount > 0 && item.historicalCount > 0) ||
+             (item.exemptCount > 0 && item.historicalCount > 0)
+    }
+
+    const noSelectedOwner = (item: ManufacturedHomeSearchResultIF): boolean => {
+      var filteredResults = localState.results?.filter(result => result.mhrNumber === item.mhrNumber)
+      filteredResults = filteredResults.filter(result => result.selected === true)
```
Comment (cameron-eyds) [inline ppr-ui/src/components/tables/mhr/SearchedResults.vue]: Can be simplified to `!!result.selected` or if this property is already of type boolean, `result.selected` would suffice.

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue (hunk 2 above)
Comment (owillborn) [reply ppr-ui/src/components/tables/mhr/SearchedResults.vue]: good call

File: ppr-ui/src/views/mhrSearch/ConfirmMHRSearch.vue [hunk 10]
```diff
@@ -195,7 +196,11 @@ export default class ConfirmMHRSearch extends Vue {
 
   private get feeQuantity (): number {
     // Return selected quantity that is not a combination search
-    return this.getSelectedManufacturedHomes.filter(result => result.selected && !result.includeLienInfo).length
+    const uniqueRegNums = []
+    this.getSelectedManufacturedHomes.forEach(result => {
+      if (!uniqueRegNums.includes(result.mhrNumber)) uniqueRegNums.push(result.mhrNumber)
+    })
+    return uniqueRegNums.length
   }
```
Comment (cameron-eyds) [inline ppr-ui/src/views/mhrSearch/ConfirmMHRSearch.vue]: This function looks like its purpose has changed.
Is it still returning selected quantity that is NOT a combination search, ie `includeLienInfo`?

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue (hunk 3 above)
Comment (owillborn) [reply ppr-ui/src/components/tables/mhr/SearchedResults.vue]: @dimak1 this is to check if ANY results have multiple names. If so we want to set the duplicate and unique classes, hiding the duplicates and disabling hover for both. Adding .length to the compare

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue (hunk 4 above)
Comment (owillborn) [reply ppr-ui/src/components/tables/mhr/SearchedResults.vue]: Looks a lot better thanks

File: ppr-ui/src/views/mhrSearch/MHRSearch.vue [hunk 11]
```diff
@@ -4,6 +4,7 @@
       <v-progress-circular color="primary" size="50" indeterminate />
     </v-overlay>
     <v-container class="container">
+      <v-icon class="mr-2" size="30" style="vertical-align: baseline; color: #212529;">mdi-home</v-icon>
```
Comment (cameron-eyds) [inline ppr-ui/src/views/mhrSearch/MHRSearch.vue]: Not ideal to put hex into the template, best practice to stick with the theme colour definitions. 

Sort of feel the same with regards to inline styling, why not create this into a local class. 
It's a small thing but keeping the template neat is always a good thing.

**Review by cameron-eyds — COMMENTED** [review]
Good job, just a few comments. 
Would you like another set of eyes on this, ie want me to pull and run it?

**Comment by owillborn** [comment]
> Good job, just a few comments. Would you like another set of eyes on this, ie want me to pull and run it?

@cameron-eyds Please do if you've got the time :)

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue (hunk 5 above)
Comment (owillborn) [reply ppr-ui/src/components/tables/mhr/SearchedResults.vue]: Need some form of sort to maintain collapsed structure. Verifying with design on the sort order

File: ppr-ui/src/views/mhrSearch/MHRSearch.vue (hunk 11 above)
Comment (cameron-eyds) [reply ppr-ui/src/views/mhrSearch/MHRSearch.vue]: To add to this, i know this was already done previously in the template. 
So no action required here, not expected you to clean up the other stuff, just sharing thoughts 👍

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue (hunk 6 above)
Comment (owillborn) [reply ppr-ui/src/components/tables/mhr/SearchedResults.vue:47]: Im not sure, but can you think of a case when uniqueResults is undefined/null? It initializes as null but is always set before use with lodashs uniqBy. If no unique selections were found it will return an empty array, meaning length of 0. Let me know if this clears things up

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue (hunk 7 above)
Comment (owillborn) [reply ppr-ui/src/components/tables/mhr/SearchedResults.vue]: Thanks! Was recently wondering the difference between var and let, this helps a lot :)

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue (hunk 6 above)
Comment (cameron-eyds) [reply ppr-ui/src/components/tables/mhr/SearchedResults.vue:47]: It does, sort of :) 

This is not a deal breaker and i purely asked because the initialized value (`null`) as you mention would result in a type error here if you tried to render this property accessor (`Cannot read property 'length' of null`) if it wasn't set or failed to set. 

In script, we can often protect ourselves from timing issues using optional chaining (`uniqueResults?.length`) but this isn't recognized in the template. 

tldr: In scenarios like this, it might be wise to initialize your property with it's intended type, to protect your template references from typeError. ie `uniqueResults = []` if safe to do so. 

IF it was initialized as `null` explicitly for some type of truthy check, you could still reference `!!localState.uniqueResults` as afaik an empty array would be falsy.

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue (hunk 5 above)
Comment (dimak1) [reply ppr-ui/src/components/tables/mhr/SearchedResults.vue]: Resolving this after discussion.

File: ppr-ui/src/interfaces/ppr-api-interfaces/manufactured-home-search-result-interface.ts [hunk 12]
```diff
@@ -4,6 +4,7 @@ import { IndividualNameIF } from '@/interfaces'
 export interface ManufacturedHomeSearchResultIF {
   id: number
   ownerName: IndividualNameIF
+  similarOwners?: Array<IndividualNameIF>
```
Comment (dimak1) [inline ppr-ui/src/interfaces/ppr-api-interfaces/manufactured-home-search-result-interface.ts]: Do we need this? I don't think we are utilizing this property.

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue (hunk 6 above)
Comment (owillborn) [reply ppr-ui/src/components/tables/mhr/SearchedResults.vue:47]: I see, I'll check with Dima if there was a purpose for init as null instead of [] as this was part of his approach

File: ppr-ui/src/views/mhrSearch/ConfirmMHRSearch.vue (hunk 10 above)
Comment (owillborn) [reply ppr-ui/src/views/mhrSearch/ConfirmMHRSearch.vue]: refactored

File: ppr-ui/src/views/mhrSearch/MHRSearch.vue (hunk 11 above)
Comment (owillborn) [reply ppr-ui/src/views/mhrSearch/MHRSearch.vue]: changes made

File: ppr-ui/src/components/tables/mhr/SearchedResults.vue (hunk 6 above)
Comment (cameron-eyds) [reply ppr-ui/src/components/tables/mhr/SearchedResults.vue:47]: Resolving the comment, this was more for awareness. Clearly it's not breaking anything 😁 
Just wanted to highlight the importance of initialization and property types and all that fun stuff 😆

File: ppr-ui/src/interfaces/ppr-api-interfaces/manufactured-home-search-result-interface.ts (hunk 12 above)
Comment (owillborn) [reply ppr-ui/src/interfaces/ppr-api-interfaces/manufactured-home-search-result-interface.ts]: removed

**Review by ChasBoucher — APPROVED** [review]
Other than comments.. Looks good.

**Review by cameron-eyds — APPROVED** [review]
Thanks for the updates, lgtm!
